### PR TITLE
Initial cut of code generator

### PR DESCRIPTION
**Pull request by myitcv**
* First cut of code generator that takes away manual aspects of creating components (i.e. where we need to handle interface requirements)
* Update to Travis scripts to include `go generate` step and final check that git is clean post tests

Remaining TODO items https://github.com/myitcv/gopherjs/wiki/Code-generator-design

Needs to be accompanied by the following update to our [Creating a GopherJS React App](https://github.com/myitcv/gopherjs/wiki/Creating-a-GopherJS-React-app) wiki:

---

### Initial setup

First verify that you can install the `reactGen` generator and it is on your `PATH`:

```bash
# only if your PATH is not already correct
export PATH=$(dirname $(go list -f '{{.Target}}' github.com/myitcv/gopherjs/cmd/reactGen)):$PATH

go install github.com/myitcv/gopherjs/cmd/reactGen
```

Then:

```bash
reactGen -help
```

should show you the options to `reactGen`, something like:

```
Usage of reactGen:
  -gglog string
        log level; one of info, warning, error, fatal (default "fatal")
  -licenseFile string
        file that contains a license header to be inserted at the top of each generated file
```

### Writing components

Let's assume we want to create a variant on the `HelloMessage` component:

```go
// File: hello_message.go
package hellomessage

import (
	"github.com/myitcv/gopherjs/react"
)

//go:generate reactGen

// Step 1
// Declare a type that has (at least) an anonymous embedded react.ComponentDef 
// (it can have other fields); this type must have the suffix 'Def', which corresponds to
// 'Definition'
//
type HelloMessageDef struct {
	react.ComponentDef
}

// Step 2
// Optionally declare a props type; the naming convention is *Props
//
type HelloMessageProps struct {
	Name string
}

// Step 3
// Optionally declare a state type; the naming convention is *State
//
type HelloMessageState struct {
	count int
}

// Step 4
// Declare a function to create instances of the component. If your component
// requires props to be specified, add this to the function signature. If the 
// props are optional, use a props pointer type.
//  
// This function must make a call to react.BlessElement with the value that is then
// returned.
//
// Convention is that this function is given the name of the component, HelloMessage
// in this instance. Because this component has props, we also accept these as part
// of the constructor.
//
func HelloMessage(p HelloMessageProps) *HelloMessageDef {
	res := &HelloMessageDef{}

	react.BlessElement(res, p)

	return res
}

// Step 5
// Define a Render method on the component's pointer type
//
func (r *HelloMessageDef) Render() Element {
	return Div(nil,
		S("Hello "+r.Props().Name),
	)
}
```

Now we run go generate (in the same directory as the component itself) to generate helper methods:

```bash
go generate
```

At this point you are done defining the component. Always re-run `go generate` after making changes to a component (you might like to setup a [watcher](https://github.com/cespare/reflex) to do just this)

You can now optionally implement:

* `ComponentWillMount()`
* `ComponentDidMount()`
* `ComponentWillReceiveProps(...)` (requires hand-spun "generated" code, see `gen_*.go` files)
* `GetInitialState() ...` (requires hand-spun "generated" code, see `gen_*.go` files)
* `ComponentWillUnmount()`

not forgetting to re-run `go generate` after any change.

See the [various examples](https://github.com/myitcv/gopherjs/wiki/Examples) for instances of these methods

### CODE REVIEW

**Comment by myitcv** [comment]
@shurcooL want to give this a look?

**Comment by shurcooL** [comment]
Sorry, I have too much on my plate at this time, can't do that. :)

File: cmd/reactGen/gen.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+			if id.Name == name+StateTypeSuffix {
+				cg.HasGetInitState = true
+				break
+			}
+
+		}
+	}
+
+	cg.pf("// Do not edit: file generated by %v\n", reactGenCmd)
```
Comment (shurcooL) [inline cmd/reactGen/gen.go]: Suggestion, consider using:

```Go
cg.pf("// Code generated by %v; DO NOT EDIT.\n", reactGenCmd)
```

That would match the currently proposed format in https://github.com/golang/go/issues/13560#issuecomment-277804473, which might end up being the final accepted one.

Comment (myitcv) [inline cmd/reactGen/gen.go]: Thanks. Done in https://github.com/myitcv/gopherjs/pull/13

**Pull request merged**